### DISCUSSION TITLE
How to use SQL editor with authenticated role to test policies with custom claims

### DISCUSSION THREAD

**Discussion by azurechen** [supabase/supabase#22482] [Questions]
I followed the guide to create an auth hook.
https://supabase.com/docs/guides/auth/custom-claims-and-role-based-access-control-rbac?language=plpgsql#create-auth-hook-to-apply-user-role

It will add a custom `user_role` attribute into the JWT token.

And I can use `auth.jwt()` in my policy to check if the user has permission to do something.
For example:

```
CREATE POLICY "Only admin can delete user profile"
ON "UserProfile"
AS RESTRICTIVE
FOR DELETE
TO authenticated
USING (auth.jwt() ->> 'user_role' = 'ADMINISTRATOR');
```

This policy restricts that only the admin can delete a user profile.

However, I cannot test this policy in the SQL editor of Supabase Studio even if I have impersonated an authenticated user.
Because the auth hook didn't trigger in the SQL editor. Therefore the `auth.jwt()` doesn't include `user_role` attribute.

<img width="1440" alt="截圖 2024-04-04 下午10 55 44" src="https://github.com/supabase/supabase/assets/1034290/b2de0e15-22e8-4955-8acb-5daba6091cb2">

the JWT token looks like
```
{"aal":"aal1","amr":[{"method":"password","timestamp":1712242279}],"aud":"authenticated","exp":1712245879,"iat":1712242279,"iss":"https://default.supabase.co/auth/v1","sub":"d4188fa0-1412-485f-9bba-137289062fb7","role":"authenticated","email":"hao@test.com","phone":null,"session_id":"749f7754-3754-4e43-a9af-30de1ef97e57","app_metadata":{"provider":"email","providers":["email"]},"user_metadata":{}}
```

If I try to delete the Profile using the following statement
```
delete from "UserProfile" where id = 'd4188fa0-1412-485f-9bba-137289062fb7'
```
It will not respect the policy to execute the delete operation.

I think that the impersonating feature provided by SQL editor allows users can test SQL statements easily.
Please tell me how can I get the right custom claim in the SQL editor. Thanks

**Comment by GaryAustin1** [comment]
I don't think the impersonation features takes auth hooks into account.  It is looking at the auth.users table to set the settings a JWT from the authorization header would have.   Because there is no jwt involved, and the authhook is only called for auth API, it does not get added.

Seems like a feature request / issue would be needed in supabase/supabase issues for the dashboard.  Not sure how hard it will be to solve.

If you can run extra SQL code you could set your claim into the settings.  
I've used this before there was impersonation: `set request.jwt.claims to '{"role":"authenticated", "sub":"5950b438-b07c-4012-8190-6ce79e4bd8e5"}';` .  That would wipe out the impersonation claims though so you would need more complicated SQL to merge in new json to request.jwt.claims.

**Reply by Keagel** [reply]
My workaround using the local development setup is to intercept the jwt in the `custom_access_token_hook` (from a real authenticated request from the front end), copy it and use `set request.jwt.claims to '...'`. Works for queries but obviously not for any direct operation on the tables using Supabase Studio.

**Comment by azurechen** [comment]
@GaryAustin1 Thanks for your reply! I think you are right.
But I still look forward there is any solution which can perfectly adopt the auth hook feature.
And I also want to ask a question to the maintainer.

https://github.com/supabase-community/supabase-custom-claims

This guide tells me how to set custom claims into the `raw_app_meta_data ` of `auth.users` table.
I think it can solve my problem because the claims will be added into the database and the SQL editor can look at it to confirm the policy.

But looks like the auth hook feature is a new style and be suggested on the official website.
Do you still suggest using `set_claims` methods to do simular things?

**Reply by GaryAustin1** [reply]
I suggest you generate an issue if you want the dashboard devs to notice.
I don't like either myself.  I would use a table and join that based on user id in RLS.  But that assumes you mainly need the claims in the database and not in the jwt on the client.  Both the custom-claims and this have a weakness that you have to force a refresh of the jwt for it to take effect.

Most people though have been using the custom-claims repository before this auth hook.   I have seen a few users though trying the auth hook and the first discovery was the claim is not reflected in the database itself.
